src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -249,53 +274,58 @@ static P11Key derivePBEKey(Token token, PBEKeySpec keySpec, String algo)
                     // the high and low parts of each char are split into 2
                     // chars. As an example, this is the transformation for
                     // a NULL terminated password "a":
-                    // char[]    =>        [   0x0061,          0x0000     ]
-                    //                          /    \           /    \
-                    // Expansion =>       [0x0000, 0x0061, 0x0000, 0x0000]
-                    //                         |       |       |       |
-                    // BMPString =>       [  0x00,   0x61,   0x00,   0x00]
+                    //   char[] password    =>  [    0x0061,         0x0000    ]
+                    //                               /    \          /    \
+                    //   char[] expPassword =>  [0x0000, 0x0061, 0x0000, 0x0000]
+                    //                               |       |       |       |
+                    //   byte[] BMPString   =>  [  0x00,   0x61,   0x00,   0x00]
                     //
                     int inputLength = (password == null) ? 0 : password.length;
-                    char[] expPassword = new char[inputLength * 2 + 2];
+                    expPassword = new char[inputLength * 2 + 2];
                     for (int i = 0, j = 0; i < inputLength; i++, j += 2) {
                         expPassword[j] = (char) ((password[i] >>> 8) & 0xFF);
                         expPassword[j + 1] = (char) (password[i] & 0xFF);
                     }
-                    password = expPassword;
                 }
                 ckMech = new CK_MECHANISM(kdfData.kdfMech,
-                        new CK_PBE_PARAMS(password, salt, itCount));
+                        new CK_PBE_PARAMS(expPassword, salt, itCount));
             }
 
-            long keyType = getKeyType(kdfData.keyAlgo);
-            CK_ATTRIBUTE[] attrs = new CK_ATTRIBUTE[
-                    switch (kdfData.op) {
-                        case ENCRYPTION, AUTHENTICATION -> 4;
-                        case GENERIC -> 5;
-                    }];
-            attrs[0] = new CK_ATTRIBUTE(CKA_CLASS, CKO_SECRET_KEY);
-            attrs[1] = new CK_ATTRIBUTE(CKA_VALUE_LEN, keySize >> 3);
-            attrs[2] = new CK_ATTRIBUTE(CKA_KEY_TYPE, keyType);
-            switch (kdfData.op) {
-                case ENCRYPTION -> attrs[3] = CK_ATTRIBUTE.ENCRYPT_TRUE;
-                case AUTHENTICATION -> attrs[3] = CK_ATTRIBUTE.SIGN_TRUE;
-                case GENERIC -> {
-                    attrs[3] = CK_ATTRIBUTE.ENCRYPT_TRUE;
-                    attrs[4] = CK_ATTRIBUTE.SIGN_TRUE;
```

Comment:
This is more or less explained in the commit message of f4b3184f302ae01be7077acd0710322a07c4aeeb. However, just in case another token (non-NSS) allows `CKK_GENERIC_SECRET` keys for symmetric encryption mechanisms (used by Cipher services), we'll still use `CKA_SIGN=true` and `CKA_ENCRYPT=true` for `PBKDF2*` derived keys.